cpp/src/reductions/scan.cu

```diff
@@ -0,0 +1,156 @@
+  {
+    rmm::device_buffer temp_storage;
+    size_t temp_storage_bytes = 0;
+    cub::DeviceScan::InclusiveScan(temp_storage.data(), temp_storage_bytes,
+                                   input, output, Op{}, size, stream);
+    // Prepare temp storage
+    temp_storage = rmm::device_buffer{temp_storage_bytes, stream, mr};
+    cub::DeviceScan::InclusiveScan(temp_storage.data(), temp_storage_bytes,
+                                   input, output, Op{}, size, stream);
+    CUDA_CHECK_LAST();
+  }
+
+  public:
+  template <typename T,
+            typename std::enable_if_t<is_supported<T>(), T>* = nullptr>
+  void operator()(const column_view& input, mutable_column_view& output,
+                  bool inclusive, 
+                  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    CUDF_EXPECTS(input.size() == output.size(),
+                 "input and output data size must be same");
+    CUDF_EXPECTS(input.type() == output.type(),
+                 "input and output data types must be same");
+
+    CUDF_EXPECTS(input.nullable() == output.nullable(),
+                 "Input column and Output column nullable mismatch (either one "
+                 "cannot be nullable)");
+
+    size_t size = input.size();
+
+    if (input.has_nulls()) {
+      auto d_input = column_device_view::create(input, stream);
+      auto it = make_null_replacement_iterator(*d_input, Op::template identity<T>());
+      if (inclusive)
+        inclusive_scan(it, output.data<T>(), size, mr, stream);
+      else
+        exclusive_scan(it, output.data<T>(), size, mr, stream);
+    } else {
+      auto it = input.data<T>();  // since scan is for arithmetic types only
+      if (inclusive)
+        inclusive_scan(it, output.data<T>(), size, mr, stream);
+      else
+        exclusive_scan(it, output.data<T>(), size, mr, stream);
+    }
+    CUDF_EXPECTS(input.null_count() == output.null_count(),
+                 "Input / output column null count mismatch");
+  }
+
+  template <typename T,
+            typename std::enable_if_t<!is_supported<T>(), T>* = nullptr>
+  void operator()(const column_view& input, mutable_column_view& output,
+                  bool inclusive, 
+                  rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    CUDF_FAIL("Non-arithmetic types not supported for `cudf::scan`");
+  }
+};
+
+std::unique_ptr<column> scan(const column_view& input,
+                             scan_op op, bool inclusive,
+                             rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                             cudaStream_t stream=0)
+{
+  CUDF_EXPECTS(is_numeric(input.type()), "Unexpected non-numeric type.");
+
+  std::unique_ptr<column> output_column = make_numeric_column(
+      input.type(), input.size(), 
+      copy_bitmask(input, stream, mr), //copy bit mask
+      input.null_count(), stream, mr);
+
+  mutable_column_view output = output_column->mutable_view();
```

Comment:
Don't do this. It's better to _return_ outputs from these functions. Otherwise it's more difficult to add string support for any of these in the future.

Reply:
moved to last level function